### PR TITLE
Blog on Pulumi Watch using Rust

### PR DESCRIPTION
**Pull request by AaronFriel**

### CODE REVIEW

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+<!--more-->
+
+When developing a web application, hot reload frameworks enable quickly iterating and confirming the
+divs are centered. When implementing new features, a test runner with a watch mode makes test-driven
+development a breeze. In languages with an interactive shell, it's natural to use that to experiment
+with new APIs. In each of these cases, engineers find value in being able to focus on writing code,
+not typing console commands.
+
+That's why [watch mode exists]({{< relref "/blog/pulumi-watch-mode-fast-inner-loop-development-for-cloud-infrastructure" >}})
```
Comment (susanev) [inline themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:30]:
```suggestion
That's why [pulumi watch mode]({{< relref "/blog/pulumi-watch-mode-fast-inner-loop-development-for-cloud-infrastructure" >}}) exists
```

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+and is one of my favorite features, and it's why I was disappointed to learn this feature [was not
+available to users on Apple Silicon Macs](https://github.com/pulumi/pulumi/pull/6497). As a Windows
```
Comment (susanev) [inline themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:32]:
```suggestion
and is one of my favorite features, and it's why I was disappointed to learn this feature was not
available to users on Apple Silicon Macs([pr disabling watch mode](https://github.com/pulumi/pulumi/pull/6497)). As a Windows
```

Comment (susanev) [reply themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:32]: though im not sure the value in linking to this, to be honest.

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+and Linux user, I discovered this while root causing a failing build on one of our providers to the
+library used to implement watch mode. Thus began a deep dive into cgo, file-watching libraries in
+Go, operating system event APIs, and finding a solution for our build issues and missing feature.
+
+To close this feature gap for one of our most important platforms, Pulumi v3.39.0 includes a binary
+built with Rust to implement file-watching. From a user perspective, “pulumi watch” works as it
+always has. And as the first component to use the Rust language, it validated Pulumi's approach to
```
Comment (susanev) [inline themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:39]:
```suggestion
built with Rust to implement file-watching. “pulumi watch” works as it
always has. And as the first component to use the Rust language, it validated Pulumi's approach to
```

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+standard deviation over 50 samples was reduced to 46ms from 216ms. It does this while being easier
+to remember and therefore use correctly, and with more terse output tuned for the use case of
+quickly iterating on changes to a stack.
+
+Watch mode has changed how I write infrastructure as code, and it brings that benefit to everyone
+without needing to install and learn third-party tools for file-watching.
+
+## The sad state of Go file-watching
+
+Returning to the cross-compilation issue, the team found a provider's builds failed after updating
```
Comment (susanev) [inline themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:80]: provider's builds is kinda difficult to read, prolly a nit.

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+had several reported issues with cross-compilation, with users reporting that it required enabling
+cgo and building on macOS to target macOS.
+
+That raised two flags. First, [cgo is not Go](https://dave.cheney.net/2016/01/18/cgo-is-not-go).
+Using cgo entails wide-ranging side effects and results in a more complex build, packaging, and
+support lifecycle. Second, continuous integration tools that support macOS runners are limited and
+typically much higher cost. It was unacceptable to allow unintentional changes in a provider or the
+Pulumi repository to impose those requirements on providers.
+
+Why did the [notify](https://github.com/rjeczalik/notify) library have these constraints? This
```
Comment (susanev) [inline themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:92]:
```suggestion
Why did the [notify library](https://github.com/rjeczalik/notify) have these constraints? This
```

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+library depended on C library support for a macOS API called FSEvents, and its Go source files
+required a complex set of build tags. While this didn't seem impossible to work around, the library
+remains unmaintained and would require us to maintain a fork to address these deficiencies. With no
+recent contributions to the library to address this or a path to removing the cgo dependency, I had
+to find another library.
+
+The strongest candidate was [fsnotify](https://github.com/fsnotify/fsnotify). This library supported
```
Comment (susanev) [inline themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:99]:
```suggestion
The strongest candidate was the [fsnotify library](https://github.com/fsnotify/fsnotify). This library supported
```

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+the platforms required, did not require cgo, and supported cross-compilation. It did lack recursive
+directory watching, though the repository listed it as a planned feature. However, in August the
+maintainer archived the repository and stepped down from maintenance. Though contributors stepped up
+to take over ownership of the repo, it left the state of the project in doubt. Without features like
+recursive file watching, the team would have to maintain a fork. The search continued.
+
+The last candidate looked at was [watcher](https://github.com/radovskyb/watcher/), which supported
```
Comment (susanev) [inline themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:106]:
```suggestion
The last candidate looked at was [watcher package](https://github.com/radovskyb/watcher/), which supported
```

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+## From library to binary and the carcinization of Pulumi
+
+One doesn't have to look hard to find CLIs that tackle this problem, many of which were at the top
+of search results while looking for libraries. Using one of these as a plugin and reading stdout for
+events instead would avoid any link-time dependency for providers. With this in mind, I considered
+building a one-off binary using the previous utilities but ruled them out. The team would become de
+facto maintainers of a fork of a complex library, in addition to needing to solve the deficiencies
+identified above. That wasn't an innovation token I could spend.
+
+I evaluated the complexity of building on [watchman](https://github.com/facebook/watchman), an
```
Comment (susanev) [inline themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:129]:
```suggestion
I evaluated the complexity of building on the [watchman service](https://github.com/facebook/watchman), an
```

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+excellent file-watching CLI open-sourced by Facebook. Unfortunately, it would be costly to maintain
+a wrapper with watchman as a dependency. While C++ isn't inaccessible, it does impose a complex
+build system and a higher barrier to entry for engineers to contribute to it. And as both a client
+and server application, it wouldn't be as simple as packaging a single binary and shelling out to
+it.
+
+Finally, I spiked on a library written in Rust, using the library underpinning the
+[watchexec](https://github.com/watchexec/watchexec) CLI. All-in, this worked out to just over 100
```
Comment (susanev) [inline themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:137]:
```suggestion
Finally, I spiked on a library written in Rust, using the library underpinning the
[watchexec CLI](https://github.com/watchexec/watchexec). All-in, this worked out to just over 100
```

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+lines of code to maintain on top of the actively maintained watchexec libraries, with a similarly
+small patch to the Pulumi CLI. And importantly, building the binary is a simple “cargo build” that
+the team can run on any platform. That implementation lives at
+[github.com/pulumi/watchutil-rs](https://github.com/pulumi/watchutil-rs) and is packaged as
+pulumi-watch, a single-purpose binary that logs on stdout when files change.
```
Comment (susanev) [inline themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:142]:
```suggestion
the team can run on any platform. The
[watchutil-rs implementation](https://github.com/pulumi/watchutil-rs) lives in GitHub and is packaged as
pulumi-watch, a single-purpose binary that logs on stdout when files change.
```

**Review by susanev — APPROVED** [review]
very well written, thx for making this happen!

kinda tore ur link text apart, hope its useful. its important for link text to be descriptive and fully understandable outside the context of the rest of the text. not only for accessibility but also for folks who scan content. happy to chat over zoom anytime if you'd like to learn more about it.

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md (hunk 2 above)
Comment (AaronFriel) [reply themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:32]: Let's nix it!

File: themes/default/content/blog/pulumi-watch-mode-with-rust/index.md (hunk 4 above)
Comment (AaronFriel) [reply themes/default/content/blog/pulumi-watch-mode-with-rust/index.md:80]: Yeah, I think I might nix "provider" here & in the earlier description or figure out a way to reword it.

Comment (AaronFriel) [reply themes/default/content/blog/pulumi-watch-mode-with-rust/index.md]: The word provider was redundant here with the following sentence, cleaned up.

**Pull request merged**